### PR TITLE
Replace mqtt with inx

### PR DESCRIPTION
**Pull request by lmoe**
# Description of change

(Dont merge yet)!

This PR introduces INX and replaces Mqtt with it. 

Docker files were updated accordingly. 

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Bug fix (a non-breaking change which fixes an issue)
- Enhancement (a non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

`make test-short`

### CODE REVIEW

File: config.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "l1": {
-    "apiAddress": "http://localhost:14265"
+    "inxAddress": "localhost:9011"
```
Comment (muXxer) [inline config.json]: our default INX port is `9029`.
private tangle points to `9011` because of the first node in the private setup.

File: packages/util/l1starter/l1starter.go [hunk 2]
```diff
@@ -21,6 +21,7 @@ type L1Starter struct {
 func New(flags *flag.FlagSet) *L1Starter {
 	s := &L1Starter{}
 	flags.StringVar(&s.Config.APIAddress, "layer1-api", "", "layer1 API address")
```
Comment (muXxer) [inline packages/util/l1starter/l1starter.go:23]: Is this still needed?

File: packages/testutil/privtangle/privtangle.go [hunk 3]
```diff
@@ -29,11 +29,11 @@ import (
 )
 
 // requires hornet, and inx plugins binaries to be in PATH
-// https://github.com/gohornet/hornet (1653bf6)
-// https://github.com/gohornet/inx-mqtt (9447a3c)
-// https://github.com/gohornet/inx-indexer (2d612d1)
-// https://github.com/gohornet/inx-coordinator (4b5c458)
-// https://github.com/gohornet/inx-faucet (77dc47e) (requires `git submodule update --init --recursive` before building )
+// https://github.com/gohornet/hornet (v2.0.0-beta.6)
```
Comment (muXxer) [inline packages/testutil/privtangle/privtangle.go]: Better use updated versions here because there was a breaking protocol change.

File: packages/parameters/parameters.go [hunk 4]
```diff
@@ -37,6 +37,7 @@ const (
 	DashboardExploreAddressURL = "dashboard.exploreAddressUrl"
 	DashboardAuth              = "dashboard.auth"
 
+	L1INXAddress   = "l1.inxAddress"
 	L1APIAddress   = "l1.apiAddress"
```
Comment (muXxer) [inline packages/parameters/parameters.go]: same here, can maybe be removed

File: packages/l1connection/l1connection.go [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+}
+
+// OutputMap implements L1Connection
+func (c *l1client) OutputMap(myAddress iotago.Address, timeout ...time.Duration) (iotago.OutputSet, error) {
+	ctxWithTimeout, cancelContext := newCtx(c.ctx, timeout...)
+	defer cancelContext()
+
+	bech32Addr := myAddress.Bech32(parameters.L1().Protocol.Bech32HRP)
+	queries := []nodeclient.IndexerQuery{
+		&nodeclient.BasicOutputsQuery{AddressBech32: bech32Addr},
```
Comment (muXxer) [inline packages/l1connection/l1connection.go:88]: you handle that these outputs could have unlock conditions like timelock etc, right?

File: packages/l1connection/l1connection.go [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+// PostTx implements L1Connection
+// sends any tx to the L1 node, then waits until the tx is confirmed.
+func (c *l1client) PostTx(tx *iotago.Transaction, timeout ...time.Duration) error {
+	ctxWithTimeout, cancelContext := newCtx(c.ctx, timeout...)
+	defer cancelContext()
+
+	// Build a Block and post it.
+	block, err := builder.NewBlockBuilder().
+		Payload(tx).
+		Tips(ctxWithTimeout, c.nodeAPIClient).
```
Comment (muXxer) [inline packages/l1connection/l1connection.go:124]: why do you need to pass tips? let the node do this!
less back and forth. Remove the local PoW and let the node do it.

File: packages/l1connection/l1connection.go [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+		Build()
+	if err != nil {
+		return fmt.Errorf("failed to build a tx: %w", err)
+	}
+	submitBlock := func(ctx context.Context, block *iotago.Block) error {
+		_, err := c.nodeAPIClient.SubmitBlock(ctx, block, parameters.L1().Protocol)
+		return err
+	}
+	err = nodeconn.DoBlockPow(ctxWithTimeout, block, c.config.UseRemotePoW, submitBlock, c.nodeAPIClient)
```
Comment (muXxer) [inline packages/l1connection/l1connection.go]: Can be removed.

File: packages/l1connection/l1connection.go [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+	}
+	c.log.Infof("Posted transaction id %v", isc.TxID(txID))
+
+	return c.waitUntilConfirmed(ctxWithTimeout, block)
+}
+
+const pollConfirmedTxInterval = 200 * time.Millisecond
+
+// waitUntilConfirmed waits until a given tx Block is confirmed, it takes care of promotions/re-attachments for that Block
+func (c *l1client) waitUntilConfirmed(ctx context.Context, block *iotago.Block) error {
```
Comment (muXxer) [inline packages/l1connection/l1connection.go:158]: IMO this whole logic needs to be changed to work correctly with more than one consensus node.
But maybe create an issue and do it in another PR.

File: packages/l1connection/l1connection.go [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+	initialAddrOutputs, err := c.OutputMap(addr)
+	if err != nil {
+		return err
+	}
+	ctxWithTimeout, cancelContext := newCtx(c.ctx, timeout...)
+	defer cancelContext()
+
+	faucetReq := fmt.Sprintf("{\"address\":%q}", addr.Bech32(parameters.L1().Protocol.Bech32HRP))
+	faucetURL := fmt.Sprintf("%s/api/enqueue", c.config.FaucetAddress)
+	httpReq, err := http.NewRequestWithContext(ctxWithTimeout, "POST", faucetURL, bytes.NewReader([]byte(faucetReq)))
```
Comment (muXxer) [inline packages/l1connection/l1connection.go]: use `http.MethodPost` instead of "POST"

**Review by jorgemmsilva — APPROVED** [review]

File: packages/util/l1starter/l1starter.go (hunk 2 above)
Comment (jorgemmsilva) [reply packages/util/l1starter/l1starter.go:23]: yes, there are more components beside wasp will still need the api address (wasp-cli, apilib, cluster tool/tests, etc)

File: packages/testutil/privtangle/privtangle.go (hunk 3 above)
Comment (jorgemmsilva) [reply packages/testutil/privtangle/privtangle.go]: thanks for pointing that out. This is updated with the latest versions we tested with, at the time those were the latest. We will test/update this with the newest stuff

**Review by ivange94 — APPROVED** [review]

File: packages/l1connection/l1connection.go [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+	if err != nil {
+		return fmt.Errorf("unable to create request: %w", err)
+	}
+	httpReq.Header.Set("Content-Type", "application/json")
+	res, err := http.DefaultClient.Do(httpReq)
+	if err != nil {
+		return fmt.Errorf("unable to call faucet: %w", err)
+	}
+	if res.StatusCode != 202 {
```
Comment (muXxer) [inline packages/l1connection/l1connection.go]: 202 is also not best practice, use http.Status something

File: packages/l1connection/l1connection.go [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+		resBody, err := io.ReadAll(res.Body)
```
Comment (muXxer) [inline packages/l1connection/l1connection.go:266]: Couldn't the body be nil in some cases?

File: packages/l1connection/l1connection.go [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+	if len(timeout) > 0 {
+		t = timeout[0]
+	}
+	return context.WithTimeout(ctx, t)
+}
+
+func setL1ProtocolParams(info *nodeclient.InfoResponse) {
+	parameters.InitL1(&parameters.L1Params{
+		// There are no limits on how big from a size perspective an essence can be, so it is just derived from 32KB - Block fields without payload = max size of the payload
+		MaxTransactionSize: 32000, // TODO should this value come from the API in the future? or some const in iotago?
```
Comment (muXxer) [inline packages/l1connection/l1connection.go:373]: 32000 ? how was this calculated?

File: packages/nodeconn/pow.go [hunk 13]
```diff
@@ -9,9 +9,45 @@ import (
...
 	"github.com/iotaledger/iota.go/v3/pow"
+	"github.com/iotaledger/wasp/packages/parameters"
 )
 
+type submitBlockFn = func(ctx context.Context, block *iotago.Block) error
+
+func DoBlockPow(ctx context.Context, block *iotago.Block, useRemotePow bool, submitBlock submitBlockFn, nodeClient *nodeclient.Client) error {
+	if useRemotePow {
+		// remote PoW: Take the Block, clear parents, clear nonce, send to node
+		block.Parents = nil
```
Comment (muXxer) [inline packages/nodeconn/pow.go:27]: why would you clear the parents? what if you want to reference a special tip?
Like in the promoteTx ?

File: packages/nodeconn/pow.go [hunk 14]
```diff
@@ -9,9 +9,45 @@ import (
...
+		block.Nonce = 0
+		err := submitBlock(ctx, block)
+		return err
+	}
+	// do the PoW
+	refreshTipsFn := func() (tips iotago.BlockIDs, err error) {
+		// refresh tips if PoW takes longer than a configured duration.
+		resp, err := nodeClient.Tips(ctx)
```
Comment (muXxer) [inline packages/nodeconn/pow.go:35]: Your refresh tips function should take care of reusing the given tips.
We have the correct logic in another inx project (I think the inx-spammer).
But to simplify it, I would let the node do the PoW.

File: packages/nodeconn/nc_chain.go [hunk 15]
```diff
@@ -145,6 +92,11 @@ func (ncc *ncChain) PullStateOutputByID(id iotago.OutputID) {
 	ncc.stateOutputHandler(id, out)
 }
 
+func shouldBeProcessed(out iotago.Output) bool {
+	// only outputs without SDRC should be processed.
+	return !out.UnlockConditionSet().HasStorageDepositReturnCondition()
```
Comment (muXxer) [inline packages/nodeconn/nc_chain.go:119]: This will be changed in the future, right?! Right?!?!??!?!??!

File: packages/nodeconn/nodeconn.go [hunk 16]
```diff
@@ -235,67 +302,58 @@ func (nc *nodeConn) GetMetrics() nodeconnmetrics.NodeConnectionMetrics {
...
-	return block, nil
+
+	return &blockID, nil
 }
 
 const pollConfirmedTxInterval = 200 * time.Millisecond
 
 // waitUntilConfirmed waits until a given tx Block is confirmed, it takes care of promotions/re-attachments for that Block
-func (nc *nodeConn) waitUntilConfirmed(ctx context.Context, block *iotago.Block) error {
+func (nc *nodeConn) waitUntilConfirmed(ctx context.Context, blockID *iotago.BlockID) error {
```
Comment (muXxer) [inline packages/nodeconn/nodeconn.go]: Same comments from my side like before in a similar function...

File: packages/l1connection/l1connection.go (hunk 6 above)
Comment (jorgemmsilva) [reply packages/l1connection/l1connection.go:124]: `l1connection` is used by tools that are to act as L1 clients, not connected via INX

File: packages/l1connection/l1connection.go (hunk 7 above)
Comment (jorgemmsilva) [reply packages/l1connection/l1connection.go]: same as https://github.com/iotaledger/wasp/pull/1264#discussion_r962626050

File: packages/l1connection/l1connection.go (hunk 8 above)
Comment (jorgemmsilva) [reply packages/l1connection/l1connection.go:158]: https://github.com/iotaledger/wasp/pull/1264#discussion_r962626050

Comment (jorgemmsilva) [reply packages/l1connection/l1connection.go:158]: we can indeed improve the logic, but its not here

File: packages/l1connection/l1connection.go (hunk 12 above)
Comment (jorgemmsilva) [reply packages/l1connection/l1connection.go:373]: as the comment above says:

> // There are no limits on how big from a size perspective an essence can be, so it is just derived from 32KB - Block fields without payload = max size of the payload

this was taken from a slack conversation, afaik there currently is no constant provided by iota.go

File: packages/nodeconn/pow.go (hunk 13 above)
Comment (jorgemmsilva) [reply packages/nodeconn/pow.go:27]: this is only used in "remotePow". I the parents were removed as per the code provided by Alex.

File: packages/nodeconn/pow.go (hunk 14 above)
Comment (jorgemmsilva) [reply packages/nodeconn/pow.go:35]: again, not every component of wasp can rely on "INX PoW". This code was taken directly from inx-spammer iirc.

File: packages/nodeconn/nc_chain.go (hunk 15 above)
Comment (jorgemmsilva) [reply packages/nodeconn/nc_chain.go:119]: >?!?!??!?!??!

less drama please.
Yes, that's the plan, but right now the focus is on making the node withstand many days without crashing.

File: packages/l1connection/l1connection.go (hunk 5 above)
Comment (jorgemmsilva) [reply packages/l1connection/l1connection.go:88]: those are handled inside each chain. We could potentially add another check for that here, but its not strictly needed

File: packages/nodeconn/nodeconn.go (hunk 16 above)
Comment (jorgemmsilva) [reply packages/nodeconn/nodeconn.go]: https://github.com/iotaledger/wasp/issues/1267

File: packages/l1connection/l1connection.go (hunk 12 above)
Comment (muXxer) [reply packages/l1connection/l1connection.go:373]: what about something like this? (payload type is not included, because its part of the payload)

```go
	payloadSizeMax := iotago.BlockBinSerializedMaxSize - // BlockSizeMax
		serializer.OneByte - // ProtocolVersion
		serializer.OneByte - // ParentCount
		(iotago.BlockMaxParents * iotago.BlockIDLength) - // Parents
		serializer.UInt32ByteSize - // PayloadLenght
		serializer.UInt64ByteSize // Nonce
```

File: packages/nodeconn/nc_chain.go (hunk 15 above)
Comment (muXxer) [reply packages/nodeconn/nc_chain.go:119]: Sorry, I was a bit shocked tbh, because all this storage deposit return stuff was added to the protocol in the first place to be able to send on chain requests without fees. Would be a shame if that is not used at the end.

File: packages/nodeconn/pow.go (hunk 13 above)
Comment (muXxer) [reply packages/nodeconn/pow.go:27]: But there are parts in your code where you rely on the given parents.
We need to talk about that and not just ignore the problem.

File: packages/l1connection/l1connection.go (hunk 6 above)
Comment (muXxer) [reply packages/l1connection/l1connection.go:124]: Still, this is not needed.
The node will add the tips, so less overhead.

File: packages/nodeconn/nc_chain.go (hunk 15 above)
Comment (jorgemmsilva) [reply packages/nodeconn/nc_chain.go:119]: I don't think that is the case. IIRC it was to keep the "protocol is feeless" and you can "send stuff for free" narratives, nothing to do with smart contracts.

Comment (jorgemmsilva) [reply packages/nodeconn/nc_chain.go:119]: anyway, it should be supported - but its not a priority IMO

File: packages/l1connection/l1connection.go (hunk 12 above)
Comment (jorgemmsilva) [reply packages/l1connection/l1connection.go:373]: thanks, will update in another PR

File: packages/nodeconn/pow.go (hunk 13 above)
Comment (jorgemmsilva) [reply packages/nodeconn/pow.go:27]: where do you see that?

File: packages/l1connection/l1connection.go (hunk 10 above)
Comment (muXxer) [reply packages/l1connection/l1connection.go]: `http.StatusAccepted`

File: packages/l1connection/l1connection.go (hunk 11 above)
Comment (jorgemmsilva) [reply packages/l1connection/l1connection.go:266]: >	// For server requests, the Request Body is always non-nil

https://pkg.go.dev/net/http#Response

**Comment by muXxer** [comment]
meeeeerge it :rocket:

**Review by muXxer — APPROVED** [review]
